org.eclipse.m2e.core/src/org/eclipse/m2e/core/embedder/IPomFacade.java

```diff
@@ -0,0 +1,25 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Christoph Läubrich
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *      Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.m2e.core.embedder;
+
+import java.io.File;
+
+import org.eclipse.core.resources.IFile;
+
+
+public interface IPomFacade {
```

Comment:
I dislike a new public undocumented API for this. Could it be made internal? (that's a rhetorical question)

Reply:
I'll add java doc for this, `IPomFacade` is needed for `IMavenProjectFacade` so we can differentiate between maven project related and "detached"  `IPomFacade` thus  `IPomFacade` needs to be public as well.

Reply:
> Do clients of m2e (eg people developing extensions) have to know about IPomFacade?

If they want to support "traditional" pom.xml and polyglot projects, yes. Beside this, IPomFacade could be used as a more lightweight method parameter in cases where you just need "the pom" (as in loading the projects object model) in contrast to the full MavenProject, e.g. as shown here for the launching case.

> Do they need to implement it?

Probably yes, if a plugin has a mean of locating the pom beside the usual polyglot/maven defaults. but we can for sure add `@noimplement` for now and see if there are strong use cases for implementation.

> explains why API addition makes everything more complex and encourages to avoid creating public API is there is no obvious value offered by consumers.

While in general this might be true, I actually noticed that this often leads to very bad API design where everyone actually *is* using internal API (best example P2), and also m2e suffer from this. Actually we use all over the place internal APIs (even in so called plugins) and that's really bad as it hinders other to write "good" plugins. 
As often there it is told we should consume our own code in platform, I think we also should consume/respect our own API to prove its useful and complete. So actually no `x-friends` no `@SupressWarning(discouraged)` should ever be necessary beside some very rare circumstances, as each of those indicates that our API is suffering from actually being useful to (outside) extenders.

Reply:
"Probably yes" is not sufficient to add an API. We need real use-cases, requirement definitions and more than 1 example to achieve good API.

> but we can for sure add @noimplement for now and see if there are strong use cases for implementation.

Or just make it internal and everything would still work as expected.

> Everyone actually is using internal API (best example P2), and also m2e suffer from this

That's the problem of consumers if they use internal APIs without contributing user-story or workforce to the project to stabilize the API. If things get broken, it's their problem. So they are not preventing the project from innovating like API addition does.
An API is rarely on 1st shot. See this "internal" phase a a maturation phase. If later there are good use-cases reported for an API and feedback and associated resource, then it can be promoted from internal to official/stable API with all the drawbacks (as it became clear there is enough value and manpower to handle the related drawbacks).

> As often there it is told we should consume our own code in platform, I think we also should consume/respect our own API to prove its useful and complete. So actually no x-friends no @SupressWarning(discouraged) should ever be necessary beside some very rare circumstances, as each of those indicates that our API is suffering from actually being useful to (outside) extenders.

While I admire this, it reminds me my first years as an Eclipse contributor. But I learned this is not as easy as you just said. APIs are a brake to innovation, they're hard to design correctly and require some real requirements to achieve such correctness. Bundle internal vs Bundle external is something that is too simplistic in practice; so there is a world in-between (exported internal modules with x-friends) which has proven to be very helpful.

Reply:
What are the use-case you maintain for this API outside of m2e? Can you please showcase how consuming this API proposal will allow eg some connectors to work better?